fontir/src/orchestration.rs

```diff
@@ -34,83 +32,306 @@ bitflags! {
         }
+    }
 
-        pub fn $has_name(&self) -> bool {
-            let id = $id;
-            self.acl.assert_read_access(&id.clone().into());
-            {
-                let rl = self.$lock_name.read();
-                if rl.is_some() {
-                    return true;
-                }
-            }
-            self.paths.target_file(&id).is_file()
+    pub fn clone_with_acl(&self, acl: Arc<AccessControlList<I>>) -> Self {
+        ContextMap {
+            acl,
+            irio: self.irio.clone(),
+            value: self.value.clone(),
         }
+    }
 
-        pub fn $setter_name(&self, value: $value_type) {
-            let id = $id;
-            self.acl.assert_write_access(&id.clone().into());
-            if self.flags.contains(Flags::EMIT_IR) {
-                let buf = $prepersist_fn(&value);
-                self.persist(&self.paths.target_file(&id), &buf);
-            }
-            set_cached(&self.$lock_name, value);
+    /// Read item that you are sure must exist. Panic if not.
+    ///
+    /// Intended for use in [Work] to access items present in
+    /// [Work::read_access]. If these are missing something is horribly
+    /// wrong and we should kerplode.
+    pub fn get(&self, id: &I) -> Arc<T> {
+        self.acl.assert_read_access(id);
+        self.value
+            .read()
+            .get(id)
+            .cloned()
+            .unwrap_or_else(|| panic!("{:?} is not available", id))
+    }
+}
+
+impl<I, T, Ir> ContextMap<I, T, Ir>
+where
+    I: Clone + Debug + PartialEq + Eq + Hash,
+    T: IdAware<I> + Persistable,
+    Ir: IrIo<I>,
+{
+    pub fn set_unconditionally(&self, value: T) {
+        let key = value.id();
+        self.acl.assert_write_access(&key);
+
+        if self.irio.emit_ir() {
+            let mut writer = self.irio.writer(&key);
+            value.write(&mut writer);
         }
-    };
+
+        self.value.write().insert(key, Arc::from(value));
+    }
 }
 
-impl Default for Flags {
-    /// Match the way gftools configures fontmake by default
-    fn default() -> Self {
-        Flags::EMIT_IR | Flags::PREFER_SIMPLE_GLYPHS
+impl<I, T, Ir> ContextMap<I, T, Ir>
+where
+    I: Clone + Debug + PartialEq + Eq + Hash,
+    T: IdAware<I> + PartialEq + Persistable,
+    Ir: IrIo<I>,
+{
+    pub fn set(&self, value: T) {
+        let key = value.id();
+        self.acl.assert_write_access(&key);
+
+        // nop?
+        if self
+            .value
+            .read()
+            .get(&key)
+            .map(|arc| **arc == value)
+            .unwrap_or(false)
+        {
+            eprintln!("set_if_changed {:?}, NO CHANGE!", key);
+            return;
+        }
+
+        eprintln!("set_if_changed {:?}, changed!", key);
+        self.set_unconditionally(value);
     }
 }
 
+pub trait IdAware<I> {
+    fn id(&self) -> I;
+}
+
+pub trait Persistable {
+    fn read(from: &mut dyn Read) -> Self;
+    fn write(&self, to: &mut dyn Write);
+}
+
+/// Reads and writes IR
```

Comment:
I'd love a bit more exposition, I assume this is going to be implemented by some type that is responsible for reading/writing... stuff? anything? specific things?

Reply:
Added